analytics.go

```diff
@@ -3,85 +3,123 @@ package analytics
 }
 
-// New client with write key.
-func New(key string) *Client {
-	c := &Client{
-		Endpoint:  Endpoint,
-		Interval:  5 * time.Second,
-		Size:      250,
-		Logger:    newDefaultLogger(),
-		Verbose:   false,
-		Transport: http.DefaultTransport,
-		key:       key,
-		msgs:      make(chan interface{}, 100),
-		quit:      make(chan struct{}),
-		shutdown:  make(chan struct{}),
-		now:       time.Now,
-		uid:       uid,
-	}
+type client struct {
+	Config
+	key string
+
+	// This channel is where the `Enqueue` method writes messages so they can be
+	// picked up and pushed by the backend goroutine taking care of applying the
+	// batching rules.
+	msgs chan interface{}
+
+	// These two channels are used to synchronize the client shutting down when
+	// `Close` is called.
+	// The first channel is closed to signal the backend goroutine that it has
+	// to stop, then the second one is closed by the backend goroutine to signal
+	// that it has finished flushing all queued messages.
+	quit     chan struct{}
+	shutdown chan struct{}
+
+	// This HTTP client is used to send requests to the backend, it uses the
+	// HTTP transport provided in the configuration.
+	http http.Client
+}
 
+// Instantiate a new client that uses the write key passed as first argument to
+// send messages to the backend.
+// The client is created with the default configuration.
+func New(writeKey string) Client {
+	// Here we can ignore the error because the default config is always valid.
+	c, _ := NewWithConfig(writeKey, Config{})
 	return c
 }
 
-func (c *Client) Enqueue(msg Message) (err error) {
-	if err = msg.validate(); err != nil {
+// Instantiate a new client that uses the write key and configuration passed as
+// arguments to send messages to the backend.
+// The function will return an error if the configuration contained impossible
+// values (like a negative flush interval for example).
+// When the function returns an error the returned client will always be nil.
+func NewWithConfig(writeKey string, config Config) (cli Client, err error) {
+	if err = config.validate(); err != nil {
 		return
 	}
-	c.once.Do(c.startLoop)
-	c.msgs <- msg.serializable(c.uid(), c.now())
+
+	c := &client{
+		Config:   makeConfig(config),
+		key:      writeKey,
+		msgs:     make(chan interface{}, 100),
+		quit:     make(chan struct{}),
+		shutdown: make(chan struct{}),
+		http: http.Client{
+			Transport: config.Transport,
+		},
+	}
+
+	go c.loop()
+
+	cli = c
 	return
 }
 
-func (c *Client) startLoop() {
-	go c.loop()
+func (c *client) Enqueue(msg Message) (err error) {
+	if err = msg.validate(); err != nil {
+		return
+	}
+
+	m := msg.serializable(c.uid(), c.now())
+
+	defer func() {
+		// When the `msgs` channel is closed writing to it will trigger a panic.
+		// To avoid letting the panic propagate to the caller we recover from it
+		// and instead report that the client has been closed and shouldn't be
+		// used anymore.
+		if recover() != nil {
+			err = io.EOF
```

Comment:
I can make this change, I usually use `EOF` to express that an I/O object has reached a state where it can't process anything anymore, plus it reuses an error that a developer is already familiar with.
I can make the change if you feel like it's more appropriate.
Should we also return `ErrClosed` in the `Close` method when it's called more than once? (we currently use `io.EOF` there as well).